Src/Particle/AMReX_ParticleContainerI.H

```diff
@@ -701,87 +722,142 @@ ParticleContainer<NStructReal, NStructInt, NArrayReal, NArrayInt, Allocator>
-                        pold.rdata(0) = static_cast<ParticleReal>(total_mass);
-                    }
-                }
+               //Ordering may make this not unique
+               for (int i = 0; i < NArrayReal; ++i)
+                 amrex::Gpu::Atomic::AddNoRet(&partData(iv,AMREX_SPACEDIM+NStructReal+i), partData(iv,AMREX_SPACEDIM)!=0.0 ? static_cast<Real>(0) : static_cast<Real>(p.rdata(NStructReal+i)));
+
+               for (int i = 0; i < AMREX_SPACEDIM; ++i)
+               {
+                   amrex::Gpu::Atomic::AddNoRet(&partData(iv,i), static_cast<Real>((p.rdata(0)*p.pos(i))));
+               }
+               for (int i = 1; i < NStructReal; ++i)
+               {
+                   amrex::Gpu::Atomic::AddNoRet(&partData(iv,AMREX_SPACEDIM+i), static_cast<Real>((p.rdata(0)*p.rdata(i))));
+               }
+               for (int i = 0; i < 1; ++i)
+                   amrex::Gpu::Atomic::AddNoRet(&partData(iv,AMREX_SPACEDIM+i), static_cast<Real>(p.rdata(0)));
             }
 
-            //
-            // Add the aggregated particles to the virtuals.
-            //
-            for (const auto& agg_particle : agg_map)
+        }); //skipping extra false argument, doing mf.setVal(0) at beginning
+
+        amrex::ParticleToMesh(*this, imf, level,
+        [=] AMREX_GPU_DEVICE (const ParticleType& p,
+                              amrex::Array4<int> const& partData,
+                              amrex::GpuArray<amrex::Real,AMREX_SPACEDIM> const& plo_loc,
+                              amrex::GpuArray<amrex::Real,AMREX_SPACEDIM> const& dxi_loc)
+        {
+
+            auto iv = getParticleCell(p, plo_loc, dxi_loc, domain);
+            if(assign_buffer_grid(iv)==-1)
             {
-                virts.push_back(agg_particle.second);
+            //if this cell has no particle id info, do a straight copy
+            if(partData(iv,0)==0)
+            {
+          //                amrex::Gpu::Atomic::AddNoRet(&partData(iv,0), VirtualParticleID);
+                amrex::Gpu::Atomic::AddNoRet(&partData(iv,0), 1);
+                for (int i = 0; i < NStructInt; ++i)
+                  amrex::Gpu::Atomic::AddNoRet(&partData(iv,1+i), p.idata(i));
+                for (int i = 0; i < NArrayInt; ++i)
+                  amrex::Gpu::Atomic::AddNoRet(&partData(iv,1+NStructInt+i), p.idata(NStructInt+i));
             }
-        }
+            }
+        });
+
+    //There may be a better way to ensure virts is the right length
+    virts.resize(imf.sum(0));
+
+    int last_offset = 0;
+    for (MFIter mfi(mf); mfi.isValid(); ++mfi)
+    {
+        const auto bx = mfi.tilebox();
+        const auto partData = mf.array(mfi);
+        const auto imf_arr = imf.array(mfi);
+
+        Gpu::DeviceVector<int> offsets(bx.numPts());
+        auto offsets_ptr = offsets.dataPtr();
+        int next_offset = Scan::ExclusiveSum((int) bx.numPts(),(imf_arr.ptr(bx.smallEnd(),0)),(offsets.dataPtr()),Scan::retSum);
+        Gpu::synchronize();
```

Comment:
I think this sync can also be removed, as ExclusiveSum should do this already.

Reply:
Removed